src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala

```diff
@@ -284,6 +284,25 @@ class VariantSampleMatrix[T](val metadata: VariantMetadata,
   def sampleVariants(fraction: Double): VariantSampleMatrix[T] =
     copy(rdd = rdd.sample(withReplacement = false, fraction, 1))
 
+  def head(n: Long): VariantSampleMatrix[T] = {
+
+    val mergeResult = (index: Int, taskResult: Int) => taskResult
+    var nVariantsPerPartition = 0
+     rdd.sparkContext.runJob(rdd,
+      (tc: TaskContext, it: Iterator[(Variant, Annotation, Iterable[T])]) => it.size,
+      Seq[Int](1),
+      (index: Int, size: Int) => nVariantsPerPartition = size
+    )
+
+    val nPartitions = math.ceil(n.toDouble/nVariantsPerPartition).toInt
+
+    copy(rdd
+        .mapPartitionsWithIndex((idx, iter) =>
+          if (idx < nPartitions) if(idx == nPartitions-1) iter.take((n % nVariantsPerPartition).toInt) else iter else Iterator())
```

Comment:
I considered take(n) but it returns an Array so I assume all data will go
to the head node...

On Saturday, June 25, 2016, Tim Poterba <notifications@github.com> wrote:

> In
> src/main/scala/org/broadinstitute/hail/variant/VariantSampleMatrix.scala
> <https://github.com/broadinstitute/hail/pull/446#discussion_r68488245>:
>
> > @@ -284,6 +284,25 @@ class VariantSampleMatrix[T](val metadata: VariantMetadata,
> >    def sampleVariants(fraction: Double): VariantSampleMatrix[T] =
> >      copy(rdd = rdd.sample(withReplacement = false, fraction, 1))
> >
> > +  def head(n: Long): VariantSampleMatrix[T] = {
> > +
> > +    val mergeResult = (index: Int, taskResult: Int) => taskResult
> > +    var nVariantsPerPartition = 0
> > +     rdd.sparkContext.runJob(rdd,
> > +      (tc: TaskContext, it: Iterator[(Variant, Annotation, Iterable[T])]) => it.size,
> > +      Seq[Int](1),
> > +      (index: Int, size: Int) => nVariantsPerPartition = size
> > +    )
> > +
> > +    val nPartitions = math.ceil(n.toDouble/nVariantsPerPartition).toInt
> > +
> > +    copy(rdd
> > +        .mapPartitionsWithIndex((idx, iter) =>
> > +          if (idx < nPartitions) if(idx == nPartitions-1) iter.take((n % nVariantsPerPartition).toInt) else iter else Iterator())
>
> This is still doing something for every partition. RDD.take(n) might be
> what we want
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/hail/pull/446/files/e2194f05fa10fb83e6e3c67a589f5da8e3a2685f#r68488245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ADVxgc3Gby77uN76RMxGhlnOG3hhPsOyks5qPSBPgaJpZM4I-Oz9>
> .
>